src/librustdoc/html/format.rs

```diff
@@ -651,44 +641,140 @@ fn generate_macro_def_id_path(
     Ok((url, ItemType::Macro, fqp))
 }
 
+fn generate_item_def_id_path(
+    def_id: DefId,
+    original_def_id: DefId,
+    cx: &Context<'_>,
+    root_path: Option<&str>,
+) -> Result<(String, ItemType, Vec<Symbol>), HrefError> {
+    use crate::rustc_trait_selection::infer::TyCtxtInferExt;
+    use crate::rustc_trait_selection::traits::query::normalize::QueryNormalizeExt;
+    use rustc_middle::traits::ObligationCause;
+
+    let tcx = cx.tcx();
+    let crate_name = tcx.crate_name(def_id.krate);
+
+    let infcx = tcx.infer_ctxt().build();
+    let def_id = infcx
+        .at(&ObligationCause::dummy(), tcx.param_env(def_id))
+        .query_normalize(ty::Binder::dummy(tcx.type_of(def_id).instantiate_identity()))
+        .map(|resolved| infcx.resolve_vars_if_possible(resolved.value))
+        .ok()
+        .and_then(|normalized| normalized.skip_binder().ty_adt_def())
+        .map(|adt| adt.did())
+        .unwrap_or(def_id);
+
+    let relative = clean::inline::item_relative_path(tcx, def_id);
+    let fqp: Vec<Symbol> = once(crate_name).chain(relative).collect();
+
+    let shortty = tcx.def_kind(def_id).into();
+    let module_fqp = to_module_fqp(shortty, &fqp);
+    let mut is_remote = false;
+
+    let url_parts = url_parts(cx.cache(), def_id, &module_fqp, &cx.current, &mut is_remote)?;
+    let (url_parts, shortty, fqp) = make_href(root_path, shortty, url_parts, &fqp, is_remote)?;
+    if def_id == original_def_id {
+        return Ok((url_parts, shortty, fqp));
+    }
+    let kind = match tcx.def_kind(original_def_id) {
+        DefKind::Variant => "variant",
+        DefKind::AssocFn => "method",
```

Comment:
Just dropping by without having any context and without having read the rest of this patch to say that you probably need to add some extra case to differentiate between `tymethod` and `method` to generate correct URLs for required associated functions (in traits).

Reply:
So the existing was not enough because for methods, it needs information about the parent (if it's a trait, then it's not a method but a `TyMethod`.